DietPi-Globals | G_RUN_CMD: Alternative output string

MichaIng opened the pull request:
https://github.com/Fourdee/DietPi/issues/1384

What do you think? Would be most beautiful, if we could give a more meaningful output, like "DietPi process settings are applied", but for me, as DietPi-Process_Tool output can be very long, it is most important, that it is shortened, if not called manually/via dietpi-launcher, as services are started/restarted quite often during installation and other maintenance tasks.

G_RUN_CMD could get an optional output by times, like `G_RUN_CMD --text "DietPi process settings are applied" /DietPi/dietpi/dietpi-process_tool 1`

Commit:
v6.0

+ DietPi-Services | Shorten output of DietPi-Process_Tool on services start

Fourdee commented:
@MichaIng 

> What do you think? Would be most beautiful, if we could give a more meaningful output, like "DietPi process settings are applied", but for me, as DietPi-Process_Tool output can be very long, it is most important, that it is shortened,

Agree 👍 

> G_RUN_CMD could get an optional output by times, like G_RUN_CMD --text "DietPi process settings are applied" /DietPi/dietpi/dietpi-process_tool 1

I've been meaning to look into this aswell ```--option``` and how to handle it correctly. 

But for now, we could add something like ```G_RUN_CMD_TEXT``` to ```G_RUN_CMD``` and export before run:
```
export G_RUN_CMD_TEXT='DietPi process settings are applied'
G_RUN_CMD /DietPi/dietpi/dietpi-process_tool 1
```

But then you get the issue of needing to ```unset G_RUN_CMD_TEXT``` afterwards. And running it inside the ```G_RUN_CMD ``` function may not reset the export from originating script (needs testing).

MichaIng commented:
@Fourdee
Looks like a simple Solution for now. Maybe I will have a look into other source codes how they handle optional arguments safely.

MichaIng commented:
@Fourdee 
Found a simple way for now, as it is possible to give "local" variables together with a command.
```
root@VM-Stretch:~# G_RUN_CMD apt update
[  OK  ] apt update
root@VM-Stretch:~# text='testing...' G_RUN_CMD apt update
[  OK  ] testing...
root@VM-Stretch:~# echo "$text"

```
Note the empty echo result, thus variable is not available in terminal/originated script.

Commit:
v6.0

DietPi-Globals | Simple alternative G_RUN_CMD text.

Commit:
v6.0

+ Use alternative text for process tool call at services (re)start

Fourdee commented:
@MichaIng 

Did a quick commit on this:
**DietPi-Process_tool | Reduction of onscreen print. Additional status print will only occur when HIERARCHY is less than 2 and/or an error occurs.**

https://github.com/Fourdee/DietPi/commit/ffa5e53e0afa5846df2e7a4515f56c1a77eea46c

![image](https://user-images.githubusercontent.com/7077026/36353877-1647ccea-14c4-11e8-9a22-e4c410b5716f.png)

![image](https://user-images.githubusercontent.com/7077026/36353916-cc8ee3da-14c4-11e8-949d-642a0b5b20c1.png)


Thoughts?

MichaIng commented:
@Fourdee 
Jep way better 👍, on >=SUB3 (e.g. dietpi-services start/restart from within dietpi-software) `DietPi-Process_Tool > Apply` only could be even enough. Would make it easier to e.g. copy&paste a whole dietpi-software installation+configuration.

Fourdee commented:
@MichaIng 

> DietPi-Process_Tool > Apply only could be even enough

Done: https://github.com/Fourdee/DietPi/commit/7ef37636a6bc9a597f6de83bcb5ce56d0452b6e5

Yep agree. I'll make the changes now before I forget 👍 

Actually, if error occurs, we will lack the software description. I'll need to rethink code.

Have to mention the program name each time when we print:
![image](https://user-images.githubusercontent.com/7077026/36355743-bfaead10-14df-11e8-9ee3-aa755e6b4b6f.png)

🈯️ 
![image](https://user-images.githubusercontent.com/7077026/36355754-eb3a30bc-14df-11e8-8a0b-40622f832d3c.png)

Commit:
v6.2

+ DietPi-Globals | G_RUN_CMD: Simplify optional alternative text + always show program+command on error.

MichaIng commented:
I adjusted this PR to just offer alternative G_RUN_CMD text, a bid simplified. This is still useful in cases, but DietPi-Process_Tool does not need change any more.

€: Further adjustments: Call variable `L_MESSAGE` to be more individual and avoid issues, if set somewhere else. `L_` to mark as intended local/script only variable.
<s>G_RUN_CMD and positive G_ERROR_HANDLER will now also show G_PROGRAM_NAME if available. Fits to other visual cleanup done with v6.2?</s> G_PROGRAM_NAME is now printed by default with G_DIETPI-NOTIFY 👍

Commit:
v6.2

+ DietPi-Services | Remove custom process tool message for now

Commit:
v6.3

tiny

Commit:
v6.3

+ Further adjustments to G_RUN_CMD optional messages

Commit:
v6.2

DietPi-Globals | G_RUN_CMD/ERROR_HANDLER: "$G_PROGRAM_NAME" is not printed by default by G_DIETPI-NOTIFY, so no need to add to G_RUN_CMD/ERROR_HANDLER output.

MichaIng commented:
@Fourdee 
Back and forth, found G_PROGRAM_NAME now printed by G_DIETPI-NOFITY by default, which is great 👍. Before, I was did not yet investigate, where exactly you implemented this. Is great that way.

Only issue I found on errors:
```
2018-02-20 12:22:04 root@micha:~# G_PROGRAM_NAME='test' G_RUN_CMD blaaa
[.     ] test | Running command: blaaa
[FAILED] test | test: blaaa
```
- The ´echo ''´ jumps to the next line, not clearing processing message line correctly.
- G_PROGRAM_NAME appears now two times, once by G_DIETPI-NOTIFY and once by G_ERROR_HANDLER.
- This PR removes both. If new line is needed, then it needs to be implemented into `G_DIETPI-NOTIFY 1 "\n$G_ERROR_HANDLER_COMMAND"`. But for me it's more beautiful, if it replaces processing message, thus is printed on same line.

So finally this method is way easier to give optional arguments/options to the function and try to read them out correctly and remove from remaining arguments, which should be executed or printed. Only danger is, if these variables are set (accidentally or by other software/system) within terminal session. Although `L_MESSAGE` should be unique enough.
But just to be sure, add some `unset L_MESSAGE` at the end of G_RUN_CMD, to prevent at least two times false message? Of course question is, if you generally agree to this method. It would make G_RUN_CMD better usable for meaningful user information, just providing the real/full command on error.

Fourdee committed:
v6.3

+ minor change: define local as lower case

Fourdee commented:
@MichaIng 

> But just to be sure, add some unset L_MESSAGE at the end of G_RUN_CMD, to prevent at least two times false message? 

Shouldn't be needed, as ```l_message``` is defined locally, it will no longer be available after function exits.

Fourdee commented:
@MichaIng 

Maybe i've misread/understood, but I dont believe ```l_message``` needs to exist?
```G_ERROR_HANDLER_COMMAND``` remains unchanged once set. So we can essentially re-use this to print info.

```G_ERROR_HANDLER_COMMAND``` is unset and re-init to null automatically with ```G_ERROR_HANDLER_RESET```, once ```G_ERROR_HANDLER``` is finished.

MichaIng commented:
@Fourdee 
I left `G_ERROR_HANDLER_COMMAND` untouched to use it on error. Otherwise error messages/github ready output would not include the exact command, executed, just the human readable explanation. But this could make it more difficult for us the investigate/for user to test the command itself separately. Also would be good to know, if e.g. parts of the command, given as variable, were expanded wrong/not available and things like this.

About `l_message`: Non global, non capital letters 👍. Yes should be not necessary to unset. Just in case the variable name is not too unique, that they could be set to environment by other software/system parts. In case of the earlier `text` this would have been a risk, but with `l_messages` we should be safe.

Fourdee commented:
@MichaIng 

I'am still failing to see why ```l_message``` is needed here instead of ```G_ERROR_HANDLER_COMMAND```. Maybe i'am simply not seeing it correctly 😉. I trust your judgement, and i'll get this merged 👍 

>  Just in case the variable name is not too unique, that they could be set to environment by other software/system parts.

```local``` will override any exported var with the same name, for the duration of the function its contained within, eg:

```
root@DietPi:~# export TEST='exported'
root@DietPi:~# echo $TEST
exported
root@DietPi:~# ./test
exported
Local
root@DietPi:~# echo $TEST
exported
```

---

source:
```
#!/bin/bash
echo $TEST

func(){

 local TEST='Local'
 echo $TEST

}

func
```

Fourdee reviewed (approved):
👍

MichaIng commented:
@Fourdee 
If `G_ERROR_HANDLER_COMMAND` would be overwritten by manual string, then the actual command string is lost for error messages: https://github.com/Fourdee/DietPi/blob/5d61613d24b21a96ffc08fc60eb8e9523e7cd8bc/dietpi/func/dietpi-globals#L422 and https://github.com/Fourdee/DietPi/blob/5d61613d24b21a96ffc08fc60eb8e9523e7cd8bc/dietpi/func/dietpi-globals#L446 I left the old command string in place to allow e.g. direct retry and see if the command itself might have been simply wrong, e.g. by wrong variable expansion.

`l_message` does not get set, if already there, thus if the variable is too common and somehow used by other software, we could see a funny command message. But I guess no problem with this for `l_message` 😉 
: https://github.com/Fourdee/DietPi/blob/5d61613d24b21a96ffc08fc60eb8e9523e7cd8bc/dietpi/func/dietpi-globals#L406
